data/modules/TradeShips/Flow.lua

```diff
@@ -0,0 +1,535 @@
+		end
+	end
+
+	-- build all suitable local routes for all ships
+	local local_routes = {}
+	for i = #ship_names, 1, -1 do
+		local id = ship_names[i]
+		local this_ship_routes = {}
+		for _,route in ipairs(routes) do
+			-- discard route if ship can't dock/land there
+			if Trader.isStarportAcceptableForShip(route.to, dummies[id]) then
+				local duration = dummies[id]:GetDurationForDistance(route.distance)
+				-- this value determines the popularity of this station among merchants
+				-- the closer it is, and the more docks in it, the more popular it is
+				local weight = math.pow(route.to.numDocks, 2) / math.pow(duration, 2)
+				table.insert(this_ship_routes, {
+					from = route.from,
+					to = route.to,
+					distance = route.distance,
+					ndocks = route.to.numDocks,
+					duration = duration,
+					weight = weight,
+				})
+			end
+		end
+		if #this_ship_routes == 0 then
+			-- no suitable local routes for this ship
+			print("WARNING: ship " .. id .. " discarded, because there are no local routes available for it")
+			table.remove(ship_names, i)
+			-- also erasing this ship from hyperspace routes
+			hyper_routes[id] = nil
+		else
+			utils.normWeights(this_ship_routes)
+			local_routes[id] = this_ship_routes
+		end
+	end
+
+	-- as a result, not a single ship might remain
+	if #ship_names == 0 then return nil end
+
+	-- now we have only those ships that can jump from somewhere and land somewhere
+	-- now set the flow on the most popular route to MAX_ROUTE_FLOW
+	-- let's collect all data on routes and calculate the most popular route weight
+	local local_routes_map = {}
+	for _,routes_for_ship in pairs(local_routes) do
+		for _, route_for_ship in ipairs(routes_for_ship) do
+			if local_routes_map[route_for_ship.to] then
+				local_routes_map[route_for_ship.to] = local_routes_map[route_for_ship.to] + route_for_ship.weight
+			else
+				local_routes_map[route_for_ship.to] = route_for_ship.weight
+			end
+		end
+	end
+	local max_route_weight = 0
+	for _, weight in pairs(local_routes_map) do
+		max_route_weight = math.max(max_route_weight, weight)
+	end
+
+	-- weight to flow conversion factor
+	local flow_per_ship = localFactors(Core.MAX_ROUTE_FLOW, system) / max_route_weight
+	-- we will reduce it, depending on the legal status of the system
+
+	local total_flow
+	-- calculate flow for local routes
+	-- and total count of ships in space
+	-- it may be necessary to recalculate the flow if the required number of
+	-- ships turns out to be too large
+	do
+		local allowed_ships = nil
+		for _ = 1, 2 do
+			local ships_in_space
+			ships_in_space = 0
+			total_flow = 0
+			for _, ship_routes in pairs(local_routes) do
+				for _, ship_route in ipairs(ship_routes) do
+					ship_route.flow = flow_per_ship * ship_route.weight
+					total_flow = total_flow + ship_route.flow
+					ship_route.amount = ship_route.flow * ship_route.duration / 3600
+					ships_in_space = ships_in_space + ship_route.amount
+				end
+			end
```

Comment:
Instead of running a loop twice and breaking halfway through, prefer to move this first half of the loop that calculates `ships_in_space` into a local function (can be defined in the local scope of the function) and call it as needed.

Reply:
Done!